servers/src/main/java/tachyon/master/TachyonMasterFaultTolerant.java

```diff
@@ -0,0 +1,126 @@
+
+import com.google.common.base.Throwables;
+
+import tachyon.Constants;
+import tachyon.LeaderSelectorClient;
+import tachyon.Version;
+import tachyon.conf.TachyonConf;
+import tachyon.master.block.BlockMaster;
+import tachyon.master.file.FileSystemMaster;
+import tachyon.master.rawtable.RawTableMaster;
+import tachyon.util.CommonUtils;
+import tachyon.util.network.NetworkAddressUtils;
+import tachyon.util.network.NetworkAddressUtils.ServiceType;
+
+/**
+ * The fault tolerant version of TachyonMaster that uses zookeeper and standby masters.
+ */
+public class TachyonMasterFaultTolerant extends TachyonMaster {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private LeaderSelectorClient mLeaderSelectorClient = null;
+
+  public TachyonMasterFaultTolerant(TachyonConf tachyonConf) {
+    super(tachyonConf);
+
+    // Set up zookeeper specific functionality.
+    try {
+      // InetSocketAddress.toString causes test issues, so build the string by hand
+      String zkName = NetworkAddressUtils.getConnectHost(ServiceType.MASTER_RPC, mTachyonConf) + ":"
+          + getMasterAddress().getPort();
+      String zkAddress = mTachyonConf.get(Constants.ZOOKEEPER_ADDRESS);
+      String zkElectionPath = mTachyonConf.get(Constants.ZOOKEEPER_ELECTION_PATH);
+      String zkLeaderPath = mTachyonConf.get(Constants.ZOOKEEPER_LEADER_PATH);
+      mLeaderSelectorClient =
+          new LeaderSelectorClient(zkAddress, zkElectionPath, zkLeaderPath, zkName);
+    } catch (Exception e) {
+      LOG.error(e.getMessage(), e);
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Start a Tachyon master server.
+   */
+  public void start() throws Exception {
+    try {
+      mLeaderSelectorClient.start();
+    } catch (IOException e) {
+      LOG.error(e.getMessage(), e);
+      throw Throwables.propagate(e);
+    }
+
+    Thread currentThread = Thread.currentThread();
+    mLeaderSelectorClient.setCurrentMasterThread(currentThread);
+    boolean started = false;
+
+    while (true) {
+      if (mLeaderSelectorClient.isLeader()) {
+        if (started) {
+          stopMasters();
+        }
+        startMasters(true);
+        started = true;
+        startServing();
+      } else {
+        // This master should be standby, and not the leader
+        if (isServing() || !started) {
+          stopServing();
+          stopMasters();
+
+          // When transitioning from master to standby, recreate the masters with a readonly
+          // journal.
+          mBlockMaster = new BlockMaster(mBlockMasterJournal.getReadOnlyJournal(), mTachyonConf);
+          mFileSystemMaster = new FileSystemMaster(mTachyonConf, mBlockMaster,
+              mFileSystemMasterJournal.getReadOnlyJournal());
+          mRawTableMaster = new RawTableMaster(mTachyonConf, mFileSystemMaster,
+              mRawTableMasterJournal.getReadOnlyJournal());
+          startMasters(false);
+          started = true;
+        }
```

Comment:
what is the else case here (e.g., `!isServing() && started`) ? shall we log error or panic?